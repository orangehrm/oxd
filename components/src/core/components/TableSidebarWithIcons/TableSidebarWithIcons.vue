<template>
  <div class="oxd-table-left-panel" :style="customStyles">
    <div v-if="headerVisible" class="header">
      <slot name="header">
        <oxd-button
            class="table-header-action-btn"
            :class="!isLeftPanelOpen ? 'no-label' : 'w-100'"
            :tooltip="!isLeftPanelOpen ? buttonData.label : null"
            flow="right"
            :label="buttonData.label"
            :iconName="buttonData.iconName"
            :size="buttonData.size"
            :style="buttonData.style"
            :displayType="buttonData.displayType"
            @click="onHeaderBtnClick"
        >
          <template v-if="buttonData.iconImageSrc" v-slot:icon>
            <img :src="buttonData.iconImageSrc" />
          </template>
        </oxd-button>
      </slot>
      <oxd-divider />
    </div>
    <div v-if="bodyVisible" class="body">
      <slot name="sidePanelBody"></slot>
    </div>
    <div v-if="listVisible" class="oxd-sidebar-footer list">
      <slot name="footer">
        <ul>
          <li
              v-for="(item, id) in sidePanelList"
              :key="id"
              @click="selectListitem(item)"
              :class="{collapsed: !isLeftPanelOpen}"
          >
            <div
                class="count-container"
                :class="{active: selectedListItem.id === item.id}"
            >
              <oxd-icon class="left-icon" :name="item.iconName" :tooltip="item.tooltip"/>
              <p v-if="isLeftPanelOpen" class="oxd-label">{{ item.label }}</p>
            </div>
          </li>
        </ul>
      </slot>
    </div>
    <oxd-icon-button
        :name="isLeftPanelOpen ? 'chevron-left' : 'chevron-right'"
        class="oxd-table-left-panel-toggle-btn"
        @click="toggleLeftPanel"
    />
  </div>
</template>

<script lang="ts">
import {defineComponent, computed, ref} from 'vue';
import Button from '@orangehrm/oxd/core/components/Button/Button.vue';
import IconButton from '@orangehrm/oxd/core/components/Button/Icon.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import Divider from '@orangehrm/oxd/core/components/Divider/Divider.vue';

export default defineComponent({
  name: 'oxd-table-icon-filter',

  components: {
    'oxd-divider': Divider,
    'oxd-icon': Icon,
    'oxd-button': Button,
    'oxd-icon-button': IconButton,
  },

  props: {
    headerVisible: {
      type: Boolean,
      default: false,
    },
    bodyVisible: {
      type: Boolean,
      default: false,
    },
    listVisible: {
      type: Boolean,
      default: false,
    },
    bubbleVisible: {
      type: Boolean,
      default: true,
    },
    button: {
      type: Object,
      default: () => null,
    },
    width: {
      type: String,
      default: '100%',
    },
    sidePanelList: {
      type: Array,
      default: () => [],
    },
    selectedListItemId: {
      type: Number,
    },
  },

  setup(props, {emit}) {
    const selectedListItem = ref<{
      id: number;
      label: string;
      active: boolean;
    }>({
      id: props.selectedListItemId,
      label: null,
      active: false,
    });
    const isLeftPanelOpen = ref<boolean>(true);

    // TODO: Optimize these duplicated methods; Sandamal
    const buttonData = computed(() => {
      const initialObject = {
        label: 'Button',
        iconName: 'oxd-add',
        iconImageSrc: null,
        size: 'long',
        displayType: 'secondary',
        style: null,
      };
      for (const key in props.button.props) {
        const value = props.button.props[key];
        if (value) {
          initialObject[key] = value;
        }
      }
      return initialObject;
    });

    const customStyles = computed(() => {
      return {
        width: isLeftPanelOpen.value ? props.width : null,
      };
    });

    const onHeaderBtnClick = () => {
      emit('sidePanelList:onHeaderBtnClick');
    };

    const selectListitem = (item: {
      id: number;
      label: string;
      active: boolean;
    }) => {
      selectedListItem.value = item;
      emit('sidePanelList:onSelect', item);
    };

    const toggleLeftPanel = () => {
      isLeftPanelOpen.value = !isLeftPanelOpen.value;
      emit('side-panel:onToggle', isLeftPanelOpen.value);
    };

    return {
      selectedListItem,
      customStyles,
      buttonData,
      isLeftPanelOpen,
      onHeaderBtnClick,
      selectListitem,
      toggleLeftPanel,
    };
  },
});
</script>

<style src="../Label/label.scss" lang="scss" scoped></style>
<style src="../TableSidebarWithIcons/table-left-panel.scss" lang="scss" scoped></style>
