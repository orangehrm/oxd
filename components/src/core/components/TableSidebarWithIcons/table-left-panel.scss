
@import '../../../styles';
@import './variables';

.oxd-table-left-panel {
  position: relative;
  border-right: 1px solid $oxd-interface-gray-lighten-4-color;
  padding-right: 10px;
  padding-left: 1rem;
  height: 100%;
  .table-left-panel-divider {
    width: calc(100% + 10px);
  }
  .body {
    margin-top: 1rem;
  }
  .oxd-table-left-panel-toggle-btn {
    position: absolute;
    width: 28px;
    height: 28px;
    top: 39px;
    right: -15px;
    font-size: 14px;
    min-width: 22px;
    min-height: 22px;
    background-color: $oxd-white-color;
    border: 1px solid $oxd-interface-gray-lighten-2-color;
    &:hover,
    &:focus {
      color: $oxd-interface-gray-darken-1-color;
    }
  }
}

.list {
  padding-top: 1.25rem !important;
  ul {
    list-style: none;
    margin: 0;
    padding: 0;
    li {
      display: flex;
      align-items: center;
      .count-container {
        display: flex;
        align-items: center;
        width: auto;
        padding: 0.5rem;
        border-radius: 1.5rem;
        width: 100%;
        .left-icon {
          height: 1.5rem;
          width: 1.5rem;
        }
        .oxd-dropdown-selected-chip {
          position: relative;
          min-width: 20px;
          width: auto;
          padding: 4px 9px;
          font-size: 0.75rem;
          font-weight: 700;
          font-family: $oxd-left-panel-font-family;
        }
        &:hover,
        &.active {
          background-color: rgba(100, 114, 140, 0.1);
          cursor: pointer;
        }
      }
      .oxd-label {
        margin: 0.25rem 0.5rem;
        font-size: 0.75rem;
        line-height: normal;
      }
      &:not(:last-child) {
        margin-bottom: 5px;
      }
      &.collapsed {
        justify-content: center;
        .count-container {
          display: table-cell;
          text-align: center;
          vertical-align: middle;
          width: auto;
        }
      }
    }
  }
}
.table-header-action-btn {
  &.no-label {
    min-width: initial;
    ::v-deep(.oxd-button-label-wrapper) {
      display: none;
    }
  }
}
