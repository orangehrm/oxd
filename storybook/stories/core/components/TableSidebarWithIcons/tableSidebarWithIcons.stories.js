import TableSidebarWithIconsStory from './TableSidebarWithIcons.story.vue';

export default {
  title: 'Example/TableSidebarWithIcons',
};
const Template = args => ({
  setup() {
    return {args};
  },
  components: {TableSidebarWithIconsStory},
  template: '<TableSidebarWithIconsStory v-bind="args" />',
});

export const Default = Template.bind({});
Default.args = {};

