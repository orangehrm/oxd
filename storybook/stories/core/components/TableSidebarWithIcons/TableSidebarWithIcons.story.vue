<template>
  <div class="container">
    <oxd-table-sidebar-with-icon
        v-if="configurations.table.leftPanel.visible"
        class="oxd-table-left-panel"
        :class="{'with-filters': configurations.table.topBar.visible}"
        width="230px"
        :side-panel-list="dataObj.stages"
        :header-visible="configurations.table.leftPanel.header.visible"
        :body-visible="configurations.table.leftPanel.body.visible"
        :list-visible="configurations.table.leftPanel.list.visible"
        :bubble-visible="configurations.table.leftPanel.list.bubble.visible"
        :button="configurations.table.leftPanel.header.button"
        :selected-list-item-id="1"
        @sidePanelList:onSelect="sidePanelListOnSelect"
        @side-panel:onToggle="toggleSidePanel"
        @sidePanelList:onHeaderBtnClick="sidePanelListOnHeaderBtnClick"
        :is-side-panel-open="isLeftPanelOpen"
    >
      <template v-slot:sidePanelBody>
        <slot name="sidePanelBody"></slot>
      </template>
      <template v-slot:footer>
        <slot name="footer"></slot>
      </template>
    </oxd-table-sidebar-with-icon>
  </div>
</template>

<script lang="ts">
import TableSidebarWithIcons from '@orangehrm/oxd/core/components/TableSidebarWithIcons/TableSidebarWithIcons.vue';
import list from './list-config.json'
import data from './list-data.json';

import {defineComponent, ref, computed} from 'vue';
export default defineComponent({
  components: {
    'oxd-table-sidebar-with-icon': TableSidebarWithIcons
  },
  setup() {
    debugger;
    const isLeftPanelOpen = ref<boolean>(false)
    const configurations = computed(() => list)
    const dataObj = computed(() => {
      return data
    })

    const sidePanelListOnSelect = item => {
      alert(`stage selected ${item}`)
    };

    const toggleSidePanel = () => {
      isLeftPanelOpen.value = !isLeftPanelOpen.value
      alert('toggled')
    }

    const sidePanelListOnHeaderBtnClick = () => {
      alert('header button clicked')
    };

    return {
      configurations,
      dataObj,
      isLeftPanelOpen,
      sidePanelListOnSelect,
      toggleSidePanel,
      sidePanelListOnHeaderBtnClick,
    }
  }
})
</script>

<style lang="scss" scoped>
</style>
