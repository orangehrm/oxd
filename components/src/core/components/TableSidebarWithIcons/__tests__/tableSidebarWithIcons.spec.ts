import {shallowMount} from '@vue/test-utils';
import TableSidebarWithIcons from '@orangehrm/oxd/core/components/TableSidebarWithIcons/TableSidebarWithIcons.vue';

describe('TableSidebarWithIcons.vue', () => {
  it('renders OXD TableSidebarWithIcons', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: true,
        bodyVisible: true,
        listVisible: true,
        selectedListItemId: '1',
        sidePanelList: [
          {
            id: -1,
            label: 'All Candidates',
            iconName: 'oxd-standard-reports',
          },
          {
            id: 1,
            label: 'Application Received',
            iconName: 'oxd-all-reports',
          },
          {
            id: 2,
            label: 'Phone Screening',
            iconName: 'oxd-my-reports',
          },
        ],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD TableSidebarWithIcons without header', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: false,
        bodyVisible: true,
        listVisible: true,
        selectedListItemId: '1',
        sidePanelList: [
          {
            id: -1,
            label: 'All Candidates',
            iconName: 'oxd-standard-reports',
          },
          {
            id: 1,
            label: 'Application Received',
            iconName: 'oxd-all-reports',
          },
          {
            id: 2,
            label: 'Phone Screening',
            iconName: 'oxd-my-reports',
          },
        ],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD TableSidebarWithIcons without body', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: true,
        bodyVisible: false,
        listVisible: true,
        selectedListItemId: '1',
        sidePanelList: [
          {
            id: -1,
            label: 'All Candidates',
            iconName: 'oxd-standard-reports',
          },
          {
            id: 1,
            label: 'Application Received',
            iconName: 'oxd-all-reports',
          },
          {
            id: 2,
            label: 'Phone Screening',
            iconName: 'oxd-my-reports',
          },
        ],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD TableSidebarWithIcons without list', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: true,
        bodyVisible: true,
        listVisible: false,
        selectedListItemId: '1',
        sidePanelList: [
          {
            id: -1,
            label: 'All Candidates',
            iconName: 'oxd-standard-reports',
          },
          {
            id: 1,
            label: 'Application Received',
            iconName: 'oxd-all-reports',
          },
          {
            id: 2,
            label: 'Phone Screening',
            iconName: 'oxd-my-reports',
          },
        ],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD TableSidebarWithIcons with diffeent item selected', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: true,
        bodyVisible: true,
        listVisible: true,
        selectedListItemId: '2',
        sidePanelList: [
          {
            id: -1,
            label: 'All Candidates',
            iconName: 'oxd-standard-reports',
          },
          {
            id: 1,
            label: 'Application Received',
            iconName: 'oxd-all-reports',
          },
          {
            id: 2,
            label: 'Phone Screening',
            iconName: 'oxd-my-reports',
          },
        ],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD TableSidebarWithIcons with empty list', () => {
    const wrapper = shallowMount(TableSidebarWithIcons, {
      props: {
        headerVisible: true,
        bodyVisible: true,
        listVisible: true,
        selectedListItemId: '2',
        sidePanelList: [],
        width: '230px',
        button: {
          props: {
            label: 'New Candidate',
            displayType: 'secondary',
            size: 'long',
            'display-type': 'label-info',
            'icon-name': 'plus',
          },
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
