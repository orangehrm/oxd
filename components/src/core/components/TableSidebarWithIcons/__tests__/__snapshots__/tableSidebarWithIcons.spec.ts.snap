// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <div class="header">
    <oxd-button-stub label="New Candidate" iconname="oxd-add" displaytype="secondary" size="long" flow="right" class="table-header-action-btn w-100"></oxd-button-stub>
    <oxd-divider-stub class="oxd-header-separator"></oxd-divider-stub>
  </div>
  <div class="body"></div>
  <div class="oxd-sidebar-footer list">
    <ul>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-standard-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">All Candidates</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-all-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Application Received</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-my-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Phone Screening</p>
        </div>
      </li>
    </ul>
  </div>
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons with diffeent item selected 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <div class="header">
    <oxd-button-stub label="New Candidate" iconname="oxd-add" displaytype="secondary" size="long" flow="right" class="table-header-action-btn w-100"></oxd-button-stub>
    <oxd-divider-stub class="oxd-header-separator"></oxd-divider-stub>
  </div>
  <div class="body"></div>
  <div class="oxd-sidebar-footer list">
    <ul>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-standard-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">All Candidates</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-all-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Application Received</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-my-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Phone Screening</p>
        </div>
      </li>
    </ul>
  </div>
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons with empty list 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <div class="header">
    <oxd-button-stub label="New Candidate" iconname="oxd-add" displaytype="secondary" size="long" flow="right" class="table-header-action-btn w-100"></oxd-button-stub>
    <oxd-divider-stub class="oxd-header-separator"></oxd-divider-stub>
  </div>
  <div class="body"></div>
  <div class="oxd-sidebar-footer list">
    <ul></ul>
  </div>
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons without body 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <div class="header">
    <oxd-button-stub label="New Candidate" iconname="oxd-add" displaytype="secondary" size="long" flow="right" class="table-header-action-btn w-100"></oxd-button-stub>
    <oxd-divider-stub class="oxd-header-separator"></oxd-divider-stub>
  </div>
  <!--v-if-->
  <div class="oxd-sidebar-footer list">
    <ul>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-standard-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">All Candidates</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-all-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Application Received</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-my-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Phone Screening</p>
        </div>
      </li>
    </ul>
  </div>
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons without header 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <!--v-if-->
  <div class="body"></div>
  <div class="oxd-sidebar-footer list">
    <ul>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-standard-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">All Candidates</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-all-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Application Received</p>
        </div>
      </li>
      <li class="">
        <div class="count-container">
          <oxd-icon-stub name="oxd-my-reports" class="left-icon"></oxd-icon-stub>
          <p class="oxd-label">Phone Screening</p>
        </div>
      </li>
    </ul>
  </div>
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;

exports[`TableSidebarWithIcons.vue renders OXD TableSidebarWithIcons without list 1`] = `
<div class="oxd-table-left-panel" style="width: 230px;">
  <div class="header">
    <oxd-button-stub label="New Candidate" iconname="oxd-add" displaytype="secondary" size="long" flow="right" class="table-header-action-btn w-100"></oxd-button-stub>
    <oxd-divider-stub class="oxd-header-separator"></oxd-divider-stub>
  </div>
  <div class="body"></div>
  <!--v-if-->
  <oxd-icon-button-stub name="chevron-left" withcontainer="true" disabled="false" flow="top" iconstyles="[object Object]" class="oxd-table-left-panel-toggle-btn"></oxd-icon-button-stub>
</div>
`;
